File: fresher_food_backend/FressFood/Models/ProductTraceability.cs
namespace FressFood.Models
{
    /// <summary>
    /// Model cho thông tin truy xuất nguồn gốc sản phẩm
    /// </summary>
    public class ProductTraceability
    {
        public string MaTruyXuat { get; set; } // Mã truy xuất duy nhất (QR Code ID)
        public string MaSanPham { get; set; } // Mã sản phẩm
        public string TenSanPham { get; set; }
        
        // Thông tin nguồn gốc
        public string NguonGoc { get; set; } // Nguồn gốc xuất xứ
        public string NhaSanXuat { get; set; } // Nhà sản xuất
        public string DiaChiSanXuat { get; set; } // Địa chỉ sản xuất
        public DateTime NgaySanXuat { get; set; } // Ngày sản xuất
        public DateTime? NgayHetHan { get; set; } // Ngày hết hạn
        
        // Thông tin vận chuyển
        public string? NhaCungCap { get; set; } // Nhà cung cấp
        public string? PhuongTienVanChuyen { get; set; } // Phương tiện vận chuyển
        public DateTime? NgayNhapKho { get; set; } // Ngày nhập kho
        
        // Thông tin chứng nhận
        public string? ChungNhanChatLuong { get; set; } // Chứng nhận chất lượng
        public string? SoChungNhan { get; set; } // Số chứng nhận
        public string? CoQuanChungNhan { get; set; } // Cơ quan chứng nhận
        
        // Thông tin blockchain
        public string? BlockchainHash { get; set; } // Hash trên blockchain
        public string? BlockchainTransactionId { get; set; } // Transaction ID trên blockchain
        public DateTime? NgayLuuBlockchain { get; set; } // Ngày lưu lên blockchain
        
        // Metadata
        public DateTime NgayTao { get; set; }
        public DateTime? NgayCapNhat { get; set; }
    }

    /// <summary>
    /// Response model khi quét QR code
    /// </summary>
    public class ProductTraceabilityResponse
    {
        public string MaTruyXuat { get; set; }
        public Product ProductInfo { get; set; }
        public ProductTraceability TraceabilityInfo { get; set; }
        public bool IsVerified { get; set; } // Đã được xác minh trên blockchain
        public string BlockchainVerificationUrl { get; set; } // Link để verify trên blockchain
    }

    /// <summary>
    /// Request model để tạo thông tin truy xuất
    /// </summary>
    public class CreateTraceabilityRequest
    {
        public string MaSanPham { get; set; }
        public string NguonGoc { get; set; }
        public string NhaSanXuat { get; set; }
        public string DiaChiSanXuat { get; set; }
        public DateTime NgaySanXuat { get; set; }
        public DateTime? NgayHetHan { get; set; }
        public string? NhaCungCap { get; set; }
        public string? PhuongTienVanChuyen { get; set; }
        public DateTime? NgayNhapKho { get; set; }
        public string? ChungNhanChatLuong { get; set; }
        public string? SoChungNhan { get; set; }
        public string? CoQuanChungNhan { get; set; }
    }
}

File: fresher_food/lib/models/ProductTraceability.dart
import 'package:fresher_food/models/Product.dart';

class ProductTraceability {
  final String maTruyXuat;
  final String maSanPham;
  final String tenSanPham;
  final String nguonGoc;
  final String nhaSanXuat;
  final String diaChiSanXuat;
  final DateTime ngaySanXuat;
  final DateTime? ngayHetHan;
  final String? nhaCungCap;
  final String? phuongTienVanChuyen;
  final DateTime? ngayNhapKho;
  final String? chungNhanChatLuong;
  final String? soChungNhan;
  final String? coQuanChungNhan;
  final String? blockchainHash;
  final String? blockchainTransactionId;
  final DateTime? ngayLuuBlockchain;
  final DateTime ngayTao;
  final DateTime? ngayCapNhat;

  ProductTraceability({
    required this.maTruyXuat,
    required this.maSanPham,
    required this.tenSanPham,
    required this.nguonGoc,
    required this.nhaSanXuat,
    required this.diaChiSanXuat,
    required this.ngaySanXuat,
    this.ngayHetHan,
    this.nhaCungCap,
    this.phuongTienVanChuyen,
    this.ngayNhapKho,
    this.chungNhanChatLuong,
    this.soChungNhan,
    this.coQuanChungNhan,
    this.blockchainHash,
    this.blockchainTransactionId,
    this.ngayLuuBlockchain,
    required this.ngayTao,
    this.ngayCapNhat,
  });

  factory ProductTraceability.fromJson(Map<String, dynamic> json) {
    return ProductTraceability(
      maTruyXuat: json['maTruyXuat']?.toString() ?? '',
      maSanPham: json['maSanPham']?.toString() ?? '',
      tenSanPham: json['tenSanPham']?.toString() ?? '',
      nguonGoc: json['nguonGoc']?.toString() ?? '',
      nhaSanXuat: json['nhaSanXuat']?.toString() ?? '',
      diaChiSanXuat: json['diaChiSanXuat']?.toString() ?? '',
      ngaySanXuat: json['ngaySanXuat'] != null
          ? DateTime.parse(json['ngaySanXuat'])
          : DateTime.now(),
      ngayHetHan: json['ngayHetHan'] != null
          ? DateTime.parse(json['ngayHetHan'])
          : null,
      nhaCungCap: json['nhaCungCap']?.toString(),
      phuongTienVanChuyen: json['phuongTienVanChuyen']?.toString(),
      ngayNhapKho: json['ngayNhapKho'] != null
          ? DateTime.parse(json['ngayNhapKho'])
          : null,
      chungNhanChatLuong: json['chungNhanChatLuong']?.toString(),
      soChungNhan: json['soChungNhan']?.toString(),
      coQuanChungNhan: json['coQuanChungNhan']?.toString(),
      blockchainHash: json['blockchainHash']?.toString(),
      blockchainTransactionId: json['blockchainTransactionId']?.toString(),
      ngayLuuBlockchain: json['ngayLuuBlockchain'] != null
          ? DateTime.parse(json['ngayLuuBlockchain'])
          : null,
      ngayTao: json['ngayTao'] != null
          ? DateTime.parse(json['ngayTao'])
          : DateTime.now(),
      ngayCapNhat: json['ngayCapNhat'] != null
          ? DateTime.parse(json['ngayCapNhat'])
          : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'maTruyXuat': maTruyXuat,
      'maSanPham': maSanPham,
      'tenSanPham': tenSanPham,
      'nguonGoc': nguonGoc,
      'nhaSanXuat': nhaSanXuat,
      'diaChiSanXuat': diaChiSanXuat,
      'ngaySanXuat': ngaySanXuat.toIso8601String(),
      'ngayHetHan': ngayHetHan?.toIso8601String(),
      'nhaCungCap': nhaCungCap,
      'phuongTienVanChuyen': phuongTienVanChuyen,
      'ngayNhapKho': ngayNhapKho?.toIso8601String(),
      'chungNhanChatLuong': chungNhanChatLuong,
      'soChungNhan': soChungNhan,
      'coQuanChungNhan': coQuanChungNhan,
      'blockchainHash': blockchainHash,
      'blockchainTransactionId': blockchainTransactionId,
      'ngayLuuBlockchain': ngayLuuBlockchain?.toIso8601String(),
      'ngayTao': ngayTao.toIso8601String(),
      'ngayCapNhat': ngayCapNhat?.toIso8601String(),
    };
  }
}

class ProductTraceabilityResponse {
  final String maTruyXuat;
  final Product? productInfo;
  final ProductTraceability traceabilityInfo;
  final bool isVerified;
  final String? blockchainVerificationUrl;

  ProductTraceabilityResponse({
    required this.maTruyXuat,
    this.productInfo,
    required this.traceabilityInfo,
    required this.isVerified,
    this.blockchainVerificationUrl,
  });

  factory ProductTraceabilityResponse.fromJson(Map<String, dynamic> json) {
    return ProductTraceabilityResponse(
      maTruyXuat: json['maTruyXuat']?.toString() ?? '',
      productInfo: json['productInfo'] != null
          ? Product.fromJson(json['productInfo'])
          : null,
      traceabilityInfo: ProductTraceability.fromJson(json['traceabilityInfo']),
      isVerified: json['isVerified'] ?? false,
      blockchainVerificationUrl: json['blockchainVerificationUrl']?.toString(),
    );
  }
}
